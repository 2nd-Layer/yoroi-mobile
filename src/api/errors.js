import ExtendableError from 'es6-error'

// thrown when the request did go through but
// backend returned an unexpected result
export class ApiError extends ExtendableError {
  constructor(request) {
    super('ApiError')
    this.request = request
  }
}

// thrown when api failed to connect to the server
export class NetworkError extends ExtendableError {
  constructor() {
    super('NetworkError')
  }
}
