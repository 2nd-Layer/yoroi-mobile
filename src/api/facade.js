// @flow
import assert from '../utils/assert'
import {TRANSACTION_STATUS} from '../types/HistoryTransaction'
import moment from 'moment'
import _ from 'lodash'
import BigNumber from 'bignumber.js'

import type {
  RawTransaction,
  TransactionStatus,
} from '../types/HistoryTransaction'

type WireTransaction = {|
  hash: string,
  inputs_address: Array<string>,
  inputs_amount: Array<string>,
  outputs_address: Array<string>,
  outputs_amount: Array<string>,
  block_num: string,
  time: string,
  tx_state: TransactionStatus,
  last_update: string,
  best_block_num: string,
|}

const facadeStatus = (status: string) => {
  const mapping = {
    Successful: TRANSACTION_STATUS.SUCCESSFUL,
    Pending: TRANSACTION_STATUS.PENDING,
    Failed: TRANSACTION_STATUS.FAILED,
  }
  assert.assert(mapping[status] != null)
  return mapping[status]
}

export const facadeTransaction = (tx: WireTransaction): RawTransaction => {
  assert.assert(
    tx.inputs_address.length === tx.inputs_amount.length,
    'Invalid data from server (inputs mismatch)',
  )
  assert.assert(
    tx.outputs_address.length === tx.outputs_amount.length,
    'Invalid data from server (outputs mismatch)',
  )

  const blockNum = tx.block_num ? parseInt(tx.block_num, 10) : null
  assert.assert(
    tx.block_num ? blockNum : true,
    `Invalid data from server (block num ${tx.block_num})`,
  )

  const bestBlockNum = parseInt(tx.best_block_num, 10)
  assert.assert(bestBlockNum, 'Invalid data from server (best block num)')

  const mapAddressAmount = ([address, amount]) => ({
    address,
    amount: new BigNumber(amount, 10),
  })

  return {
    id: tx.hash,
    status: facadeStatus(tx.tx_state),
    inputs: _.zip(tx.inputs_address, tx.inputs_amount).map(mapAddressAmount),
    outputs: _.zip(tx.outputs_address, tx.outputs_amount).map(mapAddressAmount),
    blockNum,
    submittedAt: moment(tx.time),
    lastUpdatedAt: moment(tx.last_update),
    bestBlockNum,
  }
}
