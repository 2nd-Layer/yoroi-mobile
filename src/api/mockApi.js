import utxoSumForAddresses from '../mockData/utxoSumForAddresses.json'
import utxoForAddresses from '../mockData/utxoForAddresses.json'
import txHistory from '../mockData/history.json'

type Addresses = Array<string>

export const setIsOnlineCallback = () => null

export const fetchNewTxHistory = (since: number, addresses: Addresses) =>
  Promise.resolve(txHistory)

export const fetchUtxoSum = (addresses: Array<string>) =>
  Promise.resolve(utxoSumForAddresses)

export const fetchUtxo = (addresses: Array<string>) =>
  Promise.resolve(utxoForAddresses)
