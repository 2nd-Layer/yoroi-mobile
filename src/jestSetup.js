// @flow
/* This module sets up Jest */
import fetch from 'node-fetch'
import {Logger, LogLevel} from './utils/logging'

import nodeUtil from 'util'

import {NativeModules} from 'react-native'

NativeModules.KeyStoreBridge = {
  REJECTION_MESSAGES: {},
}

// $FlowFixMe
global.TextEncoder = nodeUtil.TextEncoder
// $FlowFixMe
global.TextDecoder = nodeUtil.TextDecoder

export default {
  setup: () => {
    global.fetch = fetch
    Logger.setLogLevel(LogLevel.Warn)
  },
  debug: () => Logger.setLogLevel(LogLevel.Debug),
}
