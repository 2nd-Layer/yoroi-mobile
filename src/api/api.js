// @flow
import _ from 'lodash'
import {BigNumber} from 'bignumber.js'

import {Logger} from '../utils/logging'
import {CONFIG} from '../config'
import {NetworkError, ApiError} from './errors'
import assert from '../utils/assert'
import {checkAndFacadeTransactionAsync} from './facade'

import type {Moment} from 'moment'
import type {Transaction, RawUtxo} from '../types/HistoryTransaction'

type Addresses = Array<string>

const _checkResponse = (response, requestPayload) => {
  if (response.status !== 200) {
    Logger.debug('Bad status code from server', response.status)
    Logger.debug('Request payload:', requestPayload)
    throw new ApiError(response)
  }
}

const _fetch = (
  path: string,
  payload: any,
  networkConfig: any = CONFIG.CARDANO,
) => {
  Logger.info(`API call: ${path}`)
  const fullPath = `${networkConfig.API_ROOT}/${path}`
  return (
    fetch(fullPath, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json; charset=utf-8',
      },
      body: JSON.stringify(payload),
    })
      // Fetch throws only for network/dns/related errors, not http statuses
      .catch((e) => {
        Logger.info(`API call ${path} failed`, e)
        /* It really is TypeError according to
        https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
        */
        if (e instanceof TypeError) {
          throw new NetworkError()
        }
        throw e
      })
      .then(async (r) => {
        Logger.info(`API call ${path} finished`)

        _checkResponse(r, payload)
        const response = await r.json()
        Logger.debug('Response:', response)
        return response
      })
  )
}

export const fetchNewTxHistory = async (
  dateFrom: Moment,
  addresses: Addresses,
  networkConfig: any = CONFIG.CARDANO,
): Promise<{isLast: boolean, transactions: Array<Transaction>}> => {
  assert.preconditionCheck(
    addresses.length <= CONFIG.API.TX_HISTORY_MAX_ADDRESSES,
    'fetchNewTxHistory: too many addresses',
  )
  const response = await _fetch(
    'txs/history',
    {
      addresses,
      dateFrom: dateFrom.toISOString(),
    },
    networkConfig,
  )

  const transactions = await Promise.all(
    response.map(checkAndFacadeTransactionAsync),
  )
  return {
    transactions,
    isLast: response.length <= CONFIG.API.TX_HISTORY_RESPONSE_LIMIT,
  }
}

export const filterUsedAddresses = async (
  addresses: Addresses,
  networkConfig: any = CONFIG.CARDANO,
): Promise<Addresses> => {
  assert.preconditionCheck(
    addresses.length <= CONFIG.API.FILTER_USED_MAX_ADDRESSES,
    'filterUsedAddresses: too many addresses',
  )
  // Take a copy in case underlying data mutates during await
  const copy = [...addresses]
  const used = await _fetch(
    'addresses/filterUsed',
    {addresses: copy},
    networkConfig,
  )
  // We need to do this so that we keep original order of addresses
  return copy.filter((addr) => used.includes(addr))
}

export const fetchUTXOsForAddresses = (
  addresses: Addresses,
  networkConfig: any = CONFIG.CARDANO,
) => {
  assert.preconditionCheck(
    addresses.length <= CONFIG.API.FETCH_UTXOS_MAX_ADDRESSES,
    'fetchNewTxHistory: too many addresses',
  )
  return _fetch('txs/utxoForAddresses', {addresses}, networkConfig)
}

export const bulkFetchUTXOsForAddresses = async (
  addresses: Array<string>,
  networkConfig: any = CONFIG.CARDANO,
): Promise<Array<RawUtxo>> => {
  const chunks = _.chunk(addresses, CONFIG.API.FETCH_UTXOS_MAX_ADDRESSES)

  const responses = await Promise.all(
    chunks.map((addrs) => fetchUTXOsForAddresses(addrs, networkConfig)),
  )
  return _.flatten(responses)
}

export const submitTransaction = (signedTx: string) => {
  return _fetch('txs/signed', {signedTx})
}

export const fetchUTXOSumForAddresses = (
  addresses: Array<string>,
  networkConfig: any = CONFIG.CARDANO,
): Promise<{sum: string}> => {
  assert.preconditionCheck(
    addresses.length <= CONFIG.API.FETCH_UTXOS_MAX_ADDRESSES,
    'fetchUTXOSumForAddresses: too many addresses',
  )
  return _fetch('txs/utxoSumForAddresses', {addresses}, networkConfig)
}

export const bulkFetchUTXOSumForAddresses = async (
  addresses: Array<string>,
  networkConfig: any = CONFIG.CARDANO,
): Promise<{fundedAddresses: Array<string>, sum: BigNumber}> => {
  const chunks = _.chunk(addresses, CONFIG.API.FETCH_UTXOS_MAX_ADDRESSES)

  const responses = await Promise.all(
    chunks.map((addrs) => fetchUTXOSumForAddresses(addrs, networkConfig)),
  )
  const sum = responses.reduce(
    (x: BigNumber, y) => x.plus(new BigNumber(y.sum || 0)),
    new BigNumber(0),
  )

  const responseUTXOAddresses = await Promise.all(
    chunks.map((addrs) => fetchUTXOsForAddresses(addrs, networkConfig)),
  )

  const fundedAddresses = _.flatten(responseUTXOAddresses).map(
    (address: any) => address.receiver,
  )

  return {
    fundedAddresses: _.uniq(fundedAddresses),
    sum,
  }
}
