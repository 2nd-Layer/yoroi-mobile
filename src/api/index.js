/* eslint-disable no-unused-vars */
import * as realApi from './api'
import * as mockApi from './mockApi'

const MOCK = false
const api = MOCK ? mockApi : realApi

export default api
