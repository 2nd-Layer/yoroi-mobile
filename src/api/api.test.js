// @flow
import jestSetup from '../jestSetup'

import moment from 'moment'

import * as api from './api'
import {ApiError} from './errors'

jestSetup.setup()

describe('History API', () => {
  it('can fetch history', async () => {
    const addresses = [
      'Ae2tdPwUPEZKAx4zt8YLTGxrhX9L6R8QPWNeefZsPgwaigWab4mEw1ECUZ7',
    ]
    const ts = moment('1970-01-01')

    // We are async
    expect.assertions(1)
    const result = await api.fetchNewTxHistory(ts, addresses)

    // $FlowFixMe it seems like toMatchSnapshot is badly typed
    expect(result.transactions[0]).toMatchSnapshot({
      best_block_num: expect.any(String),
    })
  })

  it('throws ApiError on bad request', async () => {
    const addresses = []
    // mock moment
    const ts = {toISOString: () => 'not-a-date'}

    // We are async
    expect.assertions(1)

    await expect(api.fetchNewTxHistory(ts, addresses)).rejects.toThrow(ApiError)
  })

  it('filters used addresses', async () => {
    const addresses = [
      'Ae2tdPwUPEZKAx4zt8YLTGxrhX9L6R8QPWNeefZsPgwaigWab4mEw1ECUZ7',
      'Ae2tdPwUPEZ8wGxWm9VbZXFJcgLeKQJWKqREVEtHXYdqsqc4bLeGqjSwrtu',
      'Ae2tdPwUPEZ6T9qZxpao8ciAgg6ahjHRq2jV45ndZ4oPXAwrTYqN9NGUPh4',
      'Ae2tdPwUPEZN7jAbQNXXGivhavp4nSsmYtCebTcnuUmXuWDXtM3bgJzugrY',
    ]
    const used = [
      'Ae2tdPwUPEZKAx4zt8YLTGxrhX9L6R8QPWNeefZsPgwaigWab4mEw1ECUZ7',
      'Ae2tdPwUPEZ8wGxWm9VbZXFJcgLeKQJWKqREVEtHXYdqsqc4bLeGqjSwrtu',
      'Ae2tdPwUPEZ6T9qZxpao8ciAgg6ahjHRq2jV45ndZ4oPXAwrTYqN9NGUPh4',
    ]

    expect.assertions(1)
    const result = await api.filterUsedAddresses(addresses)
    expect(result).toEqual(used)
  })

  it('keeps order in filterUsedAddresses', async () => {
    const addresses = [
      'Ae2tdPwUPEZKAx4zt8YLTGxrhX9L6R8QPWNeefZsPgwaigWab4mEw1ECUZ7',
      'Ae2tdPwUPEZ8uXdGbucR1mByMcCDqwheTziFH9S3hPXJU741K6NprZ3jKFJ',
      'Ae2tdPwUPEZAyT9PgRp751Gv8UaEzaeSuNQzDC2nZzvukFBHyLEZ9usP4YR',
    ]

    expect.assertions(2)

    const result1 = await api.filterUsedAddresses(addresses)
    expect(result1).toEqual(addresses)

    addresses.reverse()

    const result2 = await api.filterUsedAddresses(addresses)
    expect(result2).toEqual(addresses)
  })
})
