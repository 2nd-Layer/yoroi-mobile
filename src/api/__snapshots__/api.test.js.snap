// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API: can fetch history 1`] = `
Object {
  "best_block_num": Any<String>,
  "block_num": "1603213",
  "hash": "58385cdaaf65bd4cd9946820485aa44646cb254b8c680ecb08ae8b8287fbf847",
  "inputs_address": Array [
    "Ae2tdPwUPEZ8tNBZxm8CtauzEi7Zay91A7nwajx4jotysBRLL6BpwzqwKa9",
    "Ae2tdPwUPEZGiayQx7JRMV9L21Be37E6Aj63Ka75hQ37vkdNnMosEPaAJXF",
  ],
  "inputs_amount": Array [
    "5000000",
    "815736577",
  ],
  "last_update": "2018-10-03T12:18:59.407Z",
  "outputs_address": Array [
    "Ae2tdPwUPEZKAx4zt8YLTGxrhX9L6R8QPWNeefZsPgwaigWab4mEw1ECUZ7",
    "Ae2tdPwUPEZA21yabP5eLgTF6UMLUKbihy2GiwW6eDNxdPevTcRp8e4AyLX",
  ],
  "outputs_amount": Array [
    "200000000",
    "620560541",
  ],
  "time": "2018-10-03T12:18:53.000Z",
  "tx_state": "Successful",
}
`;
